airflow/models/serialized_dag.py

```diff
@@ -0,0 +1,143 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Serialzed DAG table in database."""
+
+import hashlib
+from typing import Any, Dict, List, Optional, TYPE_CHECKING
+from sqlalchemy import Column, Index, Integer, String, Text, and_
+from sqlalchemy.sql import exists
+
+from airflow.models.base import Base, ID_LEN
+from airflow.utils import timezone
+from airflow.utils.db import provide_session
+from airflow.utils.sqlalchemy import UtcDateTime
+
+
+if TYPE_CHECKING:
+    from airflow.dag.serialization.serialized_dag import SerializedDAG  # noqa: F401, E501; # pylint: disable=cyclic-import
+    from airflow.models import DAG  # noqa: F401; # pylint: disable=cyclic-import
+
+
+class SerializedDagModel(Base):
+    """A database table for serialized DAGs."""
+
+    __tablename__ = 'serialized_dag'
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    fileloc = Column(String(2000))
+    # The max length of fileloc exceeds the limit of indexing.
+    fileloc_hash = Column(Integer)
+    data = Column(Text)
+    last_updated = Column(UtcDateTime)
+
+    __table_args__ = (
+        Index('idx_fileloc_hash', fileloc_hash, unique=False),
+    )
+
+    def __init__(self, dag):
+        from airflow.dag.serialization import Serialization
+
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = SerializedDagModel.dag_fileloc_hash(self.fileloc)
+        self.data = Serialization.to_json(dag)
+        self.last_updated = timezone.utcnow()
+
+    @staticmethod
+    def dag_fileloc_hash(full_filepath: str) -> int:
+        """"Hashing file location for indexing.
+
+        :param full_filepath: full filepath of DAG file
+        :return: hashed full_filepath
+        """
+        # Truncates hash to 4 bytes.
+        # TODO(coufon): hashing is needed because the length of fileloc is 2000 as
+        # an Airflow convention, which is over the limit of indexing. If we can
+        return int(0xFFFF & int(
+            hashlib.sha1(full_filepath.encode('utf-8')).hexdigest(), 16))
+
+    @classmethod
+    @provide_session
+    def write_dag(cls, dag: 'DAG', min_update_interval: Optional[int] = None, session=None):
+        """Serializes a DAG and writes it into database.
+
+        :param dag: a DAG to be written into database
+        :param min_update_interval: minimal interval in seconds to update serialized DAG
+        """
+        if min_update_interval is not None:
+            result = session.query(cls.last_updated).filter(
+                cls.dag_id == dag.dag_id).first()
+            if result is not None and (
+                    timezone.utcnow() - result.last_updated).total_seconds() < min_update_interval:
+                return
+        session.merge(cls(dag))
+        session.commit()
```

Comment:
Should we do a `session.rollback()` in cased of encountering an exception?

Reply:
A handy func you can use is:
https://github.com/apache/airflow/blob/d5ad0761fd0b33cb89258ff6924c608c3e086680/airflow/utils/db.py#L33-L45

Reply:
Yes I think it is a best practice. I will add that.

Reply:
@coufon - lets use that `create_session()` function instead of adding rolling back. May be you agreed to that only but just in-case

Reply:
Done.